### PR TITLE
refactor(jsonrpc): Initial refactoring for Structured Error handling for JSON RPC and ViewClient

### PR DESCRIPTION
**Pull request by khorolets**
Start of #3788 
It's just a beginning of refactoring. @frol and I decided that having error structures for each handler would be better solution. This is a first try with `GetBlock` handler, which doesn't change the response from jsonrpc.

To start it, I've done:
* Extract types from `chain/client` to separate crate `client-primitives`
* Extract `messages.rs` from `chain/jsonrpc/client` to separate `jsonrpc-client-primitives` crate
* Refactor `GetBlock` handler to have `GetBlockError`

Work in progress.

@frol I'd like to know what should be changed in `GetBlockError` to make it better

P.S. Commented code in the commit will be cleared while I move further.

### CODE REVIEW

File: chain/client-primitives/src/types.rs
Comment (frol) [inline chain/client-primitives/src/types.rs:157]: This is the first iteration to make sure it is compilable, right?

File: chain/client-primitives/Cargo.toml
Comment (frol) [inline chain/client-primitives/Cargo.toml:18]: Ideally, we don't want to depend on non-primitive dependencies.

Can you review the usage and consider moving certain parts around (e.g. if we moved something unnecessary to the *-primitives, and would be better to leave it in the original place)?

File: chain/jsonrpc-primitives/src/blocks.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/blocks.rs:32]: This metric should only be triggered on "Unexpected" variant. Also, we need to log it to be able to investigate the specific case since just a number is not useful (unless we confirm that it is always zero).

File: chain/jsonrpc-primitives/src/blocks.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/blocks.rs:12]: I suggest the following renaming:

* `Unexpected` -> `Unreachable` - this is the common term in Rust, and we should indicate in the error message that we would love to be informed about the error if it happens (e.g. if it only happens on someone else's instance, where we don't collect metrics), so the error message could be something along these lines: "It is a bug if you receive this error type, please, report this incident: https://github.com/near/nearcore/issues/new/choose"
* `TooManyRequests` -> `InternalError` with a help message: "The node reached its limits. Try again later. More details: {0}"

File: chain/jsonrpc-primitives/src/message.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/message.rs:158]: Do we need it?

File: chain/jsonrpc-primitives/src/metrics.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/metrics.rs:7]: note: `unreachable`

File: chain/jsonrpc/src/lib.rs
Comment (frol) [inline chain/jsonrpc/src/lib.rs:663]:
```suggestion
        let block = match result {
            Ok(block) => block,
            Err(err) => {
                return Err(RpcError::from(RpcBlockError::from(err)));
            }
        };
```

File: chain/jsonrpc/src/lib.rs
Comment (frol) [inline chain/jsonrpc/src/lib.rs:683]: ^

File: core/primitives/src/rpc.rs
Comment (frol) [inline core/primitives/src/rpc.rs:19]: Let's extract this file from core primitives and merge it into `near-jsonrpc-primitives`

File: chain/jsonrpc/src/lib.rs
Comment (frol) [inline chain/jsonrpc/src/lib.rs:633]: And the final step would be to use `RpcBlockResponse`, which will wrap `BlockView` (flatten). :pray:

File: chain/client-primitives/src/types.rs
Comment (frol) [inline chain/client-primitives/src/types.rs:175]:
```suggestion
```

**Review by frol — CHANGES_REQUESTED** [review]

File: chain/jsonrpc-primitives/src/message.rs
Comment (khorolets) [reply chain/jsonrpc-primitives/src/message.rs:158]: For a while. Will be possible to delete after all refactoring.

File: chain/jsonrpc-primitives/src/blocks.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/blocks.rs:37]: This should be just `InternalError`. I suggest we use `String` inside the `InternalError` since there is no good in storing actix error inside anyway.

File: chain/jsonrpc-primitives/src/metrics.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/metrics.rs:8]: unreachable

File: core/primitives/src/lib.rs
Comment (frol) [inline core/primitives/src/lib.rs:5]: Why is this change?

**Review by frol — CHANGES_REQUESTED** [review]

**Review by frol — APPROVED** [review]
I unblock this PR. The minor changes are agreed to be addressed, so other code owners need to review.

Review hits:

* The changes mostly target `chain/jsonrpc` (@frol and @khorolets are code-owners, so we review that)
* The rest of the changes are about extracting modules into separate crates to avoid fat dependencies, fat primitives, and circular dependencies

File: chain/jsonrpc-primitives/src/types/blocks.rs
Comment (frol) [inline chain/jsonrpc-primitives/src/types/blocks.rs:15]: Sorry for getting here once again, but let's also clarify why it is a bug, and why it should be unreachable.

```suggestion
    // NOTE: Currently, the underlying errors are too broad, and while we tried to handle
    // expected cases, we cannot statically guarantee that no other errors will be returned
    // in the future.
    // TODO #3851: Remove this variant once we can exhaustively match all the underlying errors
    #[error("It is a bug if you receive this error type, please, report this incident: https://github.com/near/nearcore/issues/new/choose. Details: {0}")]
    Unreachable(String),
```

**Review by frol — APPROVED** [review]

**Review by frol — APPROVED** [review]

**Comment by khorolets** [comment]
@bowenwang1996 could you please take a look at this PR? Thank you.

File: chain/client-primitives/src/types.rs
Comment (bowenwang1996) [inline chain/client-primitives/src/types.rs:164]: Why do we need `Unexpected`?

File: chain/jsonrpc/src/lib.rs
Comment (bowenwang1996) [inline chain/jsonrpc/src/lib.rs:631]: why is this method refactored but not other methods?

**Review by bowenwang1996 — APPROVED** [review]

File: chain/client-primitives/src/types.rs
Comment (frol) [reply chain/client-primitives/src/types.rs:164]: @khorolets let's rename it to `Unreachable` as well and add the comment as we did for the `RpcBlockError::Unreachable`.

File: chain/jsonrpc/src/lib.rs
Comment (frol) [reply chain/jsonrpc/src/lib.rs:631]: @bowenwang1996 It is to minimize the scope. I asked @khorolets to start with a single method to justify the refactoring (extracting parts into separate *-primitives crates), and then we can purely focus on jsonrpc refactoring scaling the work of refactoring the methods and also paving the road for getting rid of `Unreachable` error states (see #3851)

File: chain/jsonrpc-primitives/src/errors.rs
Comment (miraclx) [inline chain/jsonrpc-primitives/src/errors.rs:65]: Any reason why this isn't `fmt::Display`?

Comment (khorolets) [reply chain/jsonrpc-primitives/src/errors.rs:65]: No reason if I recall correct

Comment (miraclx) [reply chain/jsonrpc-primitives/src/errors.rs:65]: Resolved in #4772

**Pull request merged**
